src/virtual-machine.js

```diff
@@ -184,6 +186,27 @@ class VirtualMachine extends EventEmitter {
         return this.fromJSON(json);
     }
 
+    /**
+     * Load a project from a Scratch 3.0 sb3 file containing a project json
+     * and all of the sound and costume files.
+     * @param {JSZip} sb3File The sb3 file representing the project to load.
+     * @return {!Promise} Promise that resolves after targets are installed.
+     */
+    loadProjectLocal (sb3File) {
+        // TODO need to handle sb2 files as well, and will possibly merge w/
+        // above function
+        return sb3File.file('project.json').async('string')
+            .then(json => {
+                // TODO look at promise documentation to do this on success,
+                // but something else on error
```

Comment:
There are two main ways to handle errors with promises:
1. If it's the sort of error that you'd throw an exception for, you can call `reject` instead of `resolve`. Handling a `reject` call happens in the second arg for `then`. Note that JS debuggers tend to treat calling `reject` the same way they treat throwing an exception, so I wouldn't choose this option unless the bad thing is worthy of exception-like behavior. Here's a condensed version:
```js
new Promise(
    // There's no spec or anything that says reject must be called with an Error object, but I find it convenient.
    (resolve, reject) => { if (bad) reject(new Error("It's bad.")); else resolve("It's good."); }
  ).then(
    result => { console.log("The good result is:", result); }, // <- note the comma here
    e => { console.log("Something bad happened:", e); }
  );
```

2. Encode the error into the `resolve` result -- I imagine you already thought of this option. :)
A compliant promise gives you only one result, but that one result can be whatever works for you. For example, it could be `null` in the case of an error, or it could be a Node.js-style `err, result` pair if you stuff those into an array or an object. Some promise libraries support a [`spread`](http://bluebirdjs.com/docs/api/spread.html) call that makes this easier, but that's not (currently) standard.

In the storage engine, I originally went with option 1 for things like HTTP 404. I later switched to option 2 and just return `null` in the case of a network error, since IMO it doesn't make sense to throw an exception as part of the "try here, then try here" pattern that the storage engine uses intentionally. There may be a few remnants of the option 1 plan in the storage engine...

Reply:
Yeah, Ray and I talked about this a little bit too (I think last week). Mostly that was a TODO comment to remind myself to take care of error handling later as the workflow gets more fleshed out/cleaned up.